[Rest-Client] ConstraintDeclarationException HV000151 When using Validation Constraints on Container Elements

### Describe the bug

When using bean validation annotation on container elements e.g. `List<@Valid User> body` in a method of a RestClient interface under certain circumstances startup fails with:

`Caused by: jakarta.validation.ConstraintDeclarationException: HV000151: A method overriding another method must not redefine the parameter constraint configuration, but method ExampleRestClient$$CDIWrapper#post8(String, List) redefines the configuration of ExampleRestClient#post8(String, List).`

**Reproducer:**
add the following class to a quarkus project (and uncommenting the example methods) which uses quarkus-rest-client.
run quarkus -> startup fails
(issue is not present in quarkus-resteasy-client)

```
@Path("/example")
@RegisterRestClient
public interface ExampleRestClient {
	// OK
	@POST
	void post1(@QueryParam("a") @Size(min = 4, max = 32) String query);

	@POST void post2(@QueryParam("a") @Size(min = 4, max = 32) String query, @Valid User body);

	@POST void post3(List<@Size(min=4, max=64) String> body);

	@POST void post4(@Valid List<@Size(min=4, max=64) String> body);

	@POST void post5(@Valid User body);

	@POST void post6(List<@Valid User> body);

	@POST void post7(@QueryParam("a") String query, List<@Valid User> body);

	// NOT OK

	// @POST void post8(@NotNull List<@Valid User> body);

	// @POST void post9(@QueryParam("a") @Size(min = 4, max = 32) String query, List<@Valid User> body);

	// @POST void post10(@Valid List<@Valid User> body);
}
```




### Expected behavior

Quarkus starts, restclient works with bean validation

### Actual behavior

ERROR [io.qua.run.Quarkus] Error running Quarkus: java.lang.ExceptionInInitializerError
        at java.base/jdk.internal.misc.Unsafe.allocateInstance(Native Method)
        at java.base/java.lang.invoke.DirectMethodHandle.allocateInstance(DirectMethodHandle.java:501)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:79)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:51)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:144)
        at io.quarkus.runner.GeneratedMain.main(Unknown Source)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at io.quarkus.runner.bootstrap.StartupActionImpl$1.run(StartupActionImpl.java:121)
        at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.RuntimeException: Failed to start quarkus
        at io.quarkus.runner.ApplicationImpl.<clinit>(Unknown Source)
        ... 10 more
Caused by: jakarta.enterprise.inject.CreationException: Error creating synthetic bean [cjc3qHBmUB1ZLkEsO6nVgOk3474]: jakarta.validation.ConstraintDeclarationException: HV000151: A method overriding another method must not redefine the parameter constraint configuration, but method ExampleRestClient$$CDIWrapper#post9(List) redefines the configuration of ExampleRestClient#post9(List).

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

Linux - 6.8.0-60-generic #63-Ubuntu SMP PREEMPT_DYNAMIC Tue Apr 15 19:04:15 UTC 2025 x86_64 x86_64 x86_64 GNU/Linux

### Output of `java -version`

 21.0.7+6-Ubuntu-0ubuntu124.04, mixed mode, sharing)openjdk version "21.0.7" 2025-04-15 OpenJDK Runtime Environment (build 21.0.7+6-Ubuntu-0ubuntu124.04) OpenJDK 64-Bit Server VM (build

### Quarkus version or git rev

3.24.3

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.9 (8e8579a9e76f7d015ee5ec7bfcdc97d260186937)

### Additional information

_No response_

--- a/extensions/resteasy-reactive/rest-client/deployment/src/main/java/io/quarkus/rest/client/reactive/deployment/RestClientReactiveProcessor.java
+++ b/extensions/resteasy-reactive/rest-client/deployment/src/main/java/io/quarkus/rest/client/reactive/deployment/RestClientReactiveProcessor.java
@@ -17,6 +17,9 @@ import static io.quarkus.rest.client.reactive.deployment.DotNames.REGISTER_PROVI
 import static io.quarkus.rest.client.reactive.deployment.DotNames.RESPONSE_EXCEPTION_MAPPER;
 import static io.quarkus.rest.client.reactive.deployment.RegisteredRestClientBuildItem.toRegisteredRestClients;
 import static java.util.Arrays.asList;
+import static org.jboss.jandex.gizmo2.Jandex2Gizmo.addAnnotation;
+import static org.jboss.jandex.gizmo2.Jandex2Gizmo.classDescOf;
+import static org.jboss.jandex.gizmo2.Jandex2Gizmo.genericTypeOf;
 import static org.jboss.resteasy.reactive.common.processor.EndpointIndexer.CDI_WRAPPER_SUFFIX;
 import static org.jboss.resteasy.reactive.common.processor.JandexUtil.isImplementorOf;
 import static org.jboss.resteasy.reactive.common.processor.ResteasyReactiveDotNames.APPLICATION;
@@ -25,6 +28,7 @@ import static org.jboss.resteasy.reactive.common.processor.ResteasyReactiveDotNa
 import static org.jboss.resteasy.reactive.common.processor.scanning.ResteasyReactiveScanner.BUILTIN_HTTP_ANNOTATIONS_TO_METHOD;
 
 import java.lang.annotation.RetentionPolicy;
+import java.lang.constant.ClassDesc;
 import java.lang.reflect.Modifier;
 import java.util.ArrayList;
 import java.util.Collection;
@@ -56,6 +60,7 @@ import org.jboss.jandex.CompositeIndex;
 import org.jboss.jandex.DotName;
 import org.jboss.jandex.IndexView;
 import org.jboss.jandex.MethodInfo;
+import org.jboss.jandex.MethodParameterInfo;
 import org.jboss.jandex.Type;
 import org.jboss.logging.Logger;
 import org.jboss.resteasy.reactive.client.api.ClientLogger;
@@ -69,6 +74,7 @@ import io.quarkus.arc.deployment.AdditionalBeanBuildItem;
 import io.quarkus.arc.deployment.BeanArchiveIndexBuildItem;
 import io.quarkus.arc.deployment.CustomScopeAnnotationsBuildItem;
 import io.quarkus.arc.deployment.GeneratedBeanBuildItem;
+import io.quarkus.arc.deployment.GeneratedBeanGizmo2Adaptor;
 import io.quarkus.arc.deployment.GeneratedBeanGizmoAdaptor;
 import io.quarkus.arc.deployment.UnremovableBeanBuildItem;
 import io.quarkus.arc.processor.BuiltinScope;
@@ -95,6 +101,13 @@ import io.quarkus.gizmo.ClassCreator;
 import io.quarkus.gizmo.MethodCreator;
 import io.quarkus.gizmo.MethodDescriptor;
 import io.quarkus.gizmo.ResultHandle;
+import io.quarkus.gizmo2.Const;
+import io.quarkus.gizmo2.Expr;
+import io.quarkus.gizmo2.Gizmo;
+import io.quarkus.gizmo2.ParamVar;
+import io.quarkus.gizmo2.desc.ConstructorDesc;
+import io.quarkus.gizmo2.desc.InterfaceMethodDesc;
+import io.quarkus.gizmo2.desc.MethodDesc;
 import io.quarkus.jaxrs.client.reactive.deployment.JaxrsClientReactiveEnricherBuildItem;
 import io.quarkus.jaxrs.client.reactive.deployment.RestClientDefaultConsumesBuildItem;
 import io.quarkus.jaxrs.client.reactive.deployment.RestClientDefaultProducesBuildItem;
@@ -552,31 +565,41 @@ class RestClientReactiveProcessor {
             }
 
             String wrapperClassName = jaxrsInterface.name().toString() + CDI_WRAPPER_SUFFIX;
-            try (ClassCreator classCreator = ClassCreator.builder()
-                    .className(wrapperClassName)
-                    .classOutput(new GeneratedBeanGizmoAdaptor(generatedBeans))
-                    .interfaces(jaxrsInterface.name().toString())
-                    .superClass(RestClientReactiveCDIWrapperBase.class)
-                    .build()) {
+            // CLASS LEVEL
+            final Optional<String> configKey = registerRestClient.getConfigKey();
 
-                // CLASS LEVEL
-                final Optional<String> configKey = registerRestClient.getConfigKey();
+            configKey.ifPresent(
+                    key -> configKeys.put(jaxrsInterface.name().toString(), key));
 
-                configKey.ifPresent(
-                        key -> configKeys.put(jaxrsInterface.name().toString(), key));
+            final ScopeInfo scope = restClientsBuildTimeConfig.getScope(capabilities, jaxrsInterface)
+                    .orElse(BuiltinScope.APPLICATION).getInfo();
 
-                final ScopeInfo scope = restClientsBuildTimeConfig.getScope(capabilities, jaxrsInterface)
-                        .orElse(BuiltinScope.APPLICATION).getInfo();
+            Optional<String> baseUri = registerRestClient.getDefaultBaseUri();
+            boolean lazyDelegate = scope.getDotName().equals(REQUEST_SCOPED)
+                    || requestedRestClientMocks.contains(jaxrsInterface.name());
+
+            final String configKeyValue = configKey.orElse(null);
+            final String baseUriValue = baseUri.orElse("");
+
+            ClassDesc interfaceClassDesc = classDescOf(jaxrsInterface);
+            ClassDesc wrapperBaseClassDesc = ClassDesc.of(RestClientReactiveCDIWrapperBase.class.getName());
+
+            Gizmo gizmo = Gizmo.create(new GeneratedBeanGizmo2Adaptor(generatedBeans));
+            gizmo.class_(wrapperClassName, cc -> {
+                cc.extends_(RestClientReactiveCDIWrapperBase.class);
+                cc.implements_(interfaceClassDesc);
+
                 // add a scope annotation, e.g. @Singleton
-                classCreator.addAnnotation(scope.getDotName().toString());
-                classCreator.addAnnotation(RestClient.class);
+                cc.addAnnotation(classDescOf(scope.getDotName()), RetentionPolicy.RUNTIME, ab -> {
+                });
+                cc.addAnnotation(RestClient.class);
+
                 // e.g. @Typed({InterfaceClass.class})
                 // needed for CDI to inject the proper wrapper in case of
                 // subinterfaces
-                org.objectweb.asm.Type asmType = org.objectweb.asm.Type
-                        .getObjectType(jaxrsInterface.name().toString().replace('.', '/'));
-                classCreator.addAnnotation(Typed.class.getName(), RetentionPolicy.RUNTIME)
-                        .addValue("value", new org.objectweb.asm.Type[] { asmType });
+                cc.addAnnotation(Typed.class, ab -> {
+                    ab.addArray("value", new ClassDesc[] { interfaceClassDesc });
+                });
 
                 for (AnnotationInstance annotation : annotationsStore.getAnnotations(jaxrsInterface)) {
                     if (SKIP_COPYING_ANNOTATIONS_TO_GENERATED_CLASS.contains(annotation.name())) {
@@ -588,28 +611,23 @@ class RestClientReactiveProcessor {
                         continue;
                     }
 
-                    classCreator.addAnnotation(annotation);
+                    addAnnotation(cc, annotation, index);
                 }
 
                 // CONSTRUCTOR:
-
-                MethodCreator constructor = classCreator
-                        .getMethodCreator(MethodDescriptor.ofConstructor(classCreator.getClassName()));
-
-                Optional<String> baseUri = registerRestClient.getDefaultBaseUri();
-
-                ResultHandle baseUriHandle = constructor.load(baseUri.isPresent() ? baseUri.get() : "");
-                boolean lazyDelegate = scope.getDotName().equals(REQUEST_SCOPED)
-                        || requestedRestClientMocks.contains(jaxrsInterface.name());
-                constructor.invokeSpecialMethod(
-                        MethodDescriptor.ofConstructor(RestClientReactiveCDIWrapperBase.class, Class.class, String.class,
-                                String.class, boolean.class),
-                        constructor.getThis(),
-                        constructor.loadClassFromTCCL(jaxrsInterface.toString()),
-                        baseUriHandle,
-                        configKey.isPresent() ? constructor.load(configKey.get()) : constructor.loadNull(),
-                        constructor.load(lazyDelegate));
-                constructor.returnValue(null);
+                cc.constructor(ctor -> {
+                    ctor.body(bc -> {
+                        ConstructorDesc superCtor = ConstructorDesc.of(wrapperBaseClassDesc,
+                                Class.class, String.class, String.class, boolean.class);
+                        bc.invokeSpecial(superCtor,
+                                cc.this_(),
+                                Const.of(interfaceClassDesc),
+                                Const.of(baseUriValue),
+                                configKeyValue != null ? Const.of(configKeyValue) : Const.ofNull(String.class),
+                                Const.of(lazyDelegate));
+                        bc.return_();
+                    });
+                });
 
                 // METHODS:
                 for (MethodInfo method : methodsToImplement) {
@@ -622,50 +640,75 @@ class RestClientReactiveProcessor {
                     // public JsonArray get() {
                     //     return InterfaceClass.super.get();
                     // }
-                    MethodCreator methodCreator = classCreator.getMethodCreator(MethodDescriptor.of(method));
-                    for (Type exception : method.exceptions()) {
-                        // declared exceptions are important in case the client is intercepted, because interception
-                        // subclasses wrap undeclared checked exceptions into `ArcUndeclaredThrowableException`
-                        methodCreator.addException(exception.name().toString());
-                    }
-                    methodCreator.setSignature(method.genericSignatureIfRequired());
+                    cc.method(method.name(), mc -> {
+                        mc.public_();
+                        mc.returning(genericTypeOf(method.returnType(), index));
 
-                    // copy method annotations, there can be interceptors bound to them:
-                    for (AnnotationInstance annotation : annotationsStore.getAnnotations(method)) {
-                        if (annotation.target().kind() == AnnotationTarget.Kind.METHOD
-                                && !BUILTIN_HTTP_ANNOTATIONS_TO_METHOD.containsKey(annotation.name())
-                                && !ResteasyReactiveDotNames.PATH.equals(annotation.name())) {
-                            methodCreator.addAnnotation(annotation);
+                        // Collect parameter annotations grouped by position
+                        Map<Short, List<AnnotationInstance>> paramAnnotations = new HashMap<>();
+                        for (AnnotationInstance annotation : annotationsStore.getAnnotations(method)) {
+                            if (annotation.target().kind() == AnnotationTarget.Kind.METHOD_PARAMETER) {
+                                short position = annotation.target().asMethodParameter().position();
+                                paramAnnotations.computeIfAbsent(position, k -> new ArrayList<>()).add(annotation);
+                            }
                         }
-                        if (annotation.target().kind() == AnnotationTarget.Kind.METHOD_PARAMETER) {
-                            // TODO should skip annotations like `@PathParam` / `@RestPath`, probably (?)
-                            short position = annotation.target().asMethodParameter().position();
-                            methodCreator.getParameterAnnotations(position).addAnnotation(annotation);
+
+                        List<ParamVar> params = new ArrayList<>();
+                        List<MethodParameterInfo> methodParams = method.parameters();
+                        for (short i = 0; i < methodParams.size(); i++) {
+                            MethodParameterInfo param = methodParams.get(i);
+                            List<AnnotationInstance> annotations = paramAnnotations.get(i);
+                            final IndexView indexView = index;
+                            params.add(mc.parameter(param.nameOrDefault(), pc -> {
+                                pc.setType(genericTypeOf(param.type(), indexView));
+                                if (annotations != null) {
+                                    for (AnnotationInstance annotation : annotations) {
+                                        addAnnotation(pc, annotation, indexView);
+                                    }
+                                }
+                            }));
                         }
-                    }
 
-                    ResultHandle result;
+                        for (Type exception : method.exceptions()) {
+                            mc.throws_(classDescOf(exception));
+                        }
 
-                    int parameterCount = method.parameterTypes().size();
-                    ResultHandle[] params = new ResultHandle[parameterCount];
-                    for (int i = 0; i < parameterCount; i++) {
-                        params[i] = methodCreator.getMethodParam(i);
-                    }
+                        // copy method annotations, there can be interceptors bound to them:
+                        for (AnnotationInstance annotation : annotationsStore.getAnnotations(method)) {
+                            if (annotation.target().kind() == AnnotationTarget.Kind.METHOD
+                                    && !BUILTIN_HTTP_ANNOTATIONS_TO_METHOD.containsKey(annotation.name())
+                                    && !ResteasyReactiveDotNames.PATH.equals(annotation.name())) {
+                                addAnnotation(mc, annotation, index);
+                            }
+                        }
 
-                    if (Modifier.isAbstract(method.flags())) { // RestClient method
-                        ResultHandle delegate = methodCreator.invokeVirtualMethod(
-                                MethodDescriptor.ofMethod(RestClientReactiveCDIWrapperBase.class, "getDelegate",
-                                        Object.class),
-                                methodCreator.getThis());
+                        mc.body(bc -> {
+                            Expr result;
+                            if (Modifier.isAbstract(method.flags())) { // RestClient method
+                                MethodDesc getDelegateMethod = MethodDesc.of(RestClientReactiveCDIWrapperBase.class,
+                                        "getDelegate",
+                                        Object.class);
+                                Expr delegate = bc.invokeVirtual(getDelegateMethod, cc.this_());
 
-                        result = methodCreator.invokeInterfaceMethod(method, delegate, params);
-                    } else { // default method
-                        result = methodCreator.invokeSpecialInterfaceMethod(method, methodCreator.getThis(), params);
-                    }
-
-                    methodCreator.returnValue(result);
+                                InterfaceMethodDesc interfaceMethod = InterfaceMethodDesc.of(
+                                        interfaceClassDesc,
+                                        method.name(),
+                                        classDescOf(method.returnType()),
+                                        method.parameterTypes().stream().map(t -> classDescOf(t)).toArray(ClassDesc[]::new));
+                                result = bc.invokeInterface(interfaceMethod, delegate, params);
+                            } else { // default method
+                                InterfaceMethodDesc interfaceMethod = InterfaceMethodDesc.of(
+                                        interfaceClassDesc,
+                                        method.name(),
+                                        classDescOf(method.returnType()),
+                                        method.parameterTypes().stream().map(t -> classDescOf(t)).toArray(ClassDesc[]::new));
+                                result = bc.invokeSpecial(interfaceMethod, cc.this_(), params);
+                            }
+                            bc.return_(result);
+                        });
+                    });
                 }
-            }
+            });
         }
 
         Set<String> blockingClassNames = new HashSet<>();

--- a/extensions/resteasy-reactive/rest-client/deployment/src/test/java/io/quarkus/rest/client/reactive/TypeParamsTest.java
+++ b/extensions/resteasy-reactive/rest-client/deployment/src/test/java/io/quarkus/rest/client/reactive/TypeParamsTest.java
@@ -1,0 +1,107 @@
+package io.quarkus.rest.client.reactive;
+
+import static org.assertj.core.api.Assertions.*;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+import java.lang.reflect.AnnotatedParameterizedType;
+import java.lang.reflect.AnnotatedType;
+import java.lang.reflect.Method;
+import java.util.List;
+import java.util.Map;
+
+import jakarta.ws.rs.POST;
+import jakarta.ws.rs.Path;
+
+import org.eclipse.microprofile.rest.client.inject.RegisterRestClient;
+import org.eclipse.microprofile.rest.client.inject.RestClient;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.arc.ClientProxy;
+import io.quarkus.test.QuarkusUnitTest;
+
+public class TypeParamsTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .withApplicationRoot((jar) -> jar.addClasses(Client.class, Pojo.class, Foo.class))
+            .overrideRuntimeConfigKey("quarkus.rest-client.client.url", "http://localhost:${quarkus.http.test-port:8081}");
+
+    @RestClient
+    Client client;
+
+    @Test
+    void test() {
+        assertThat(ClientProxy.unwrap(client).getClass()).hasMethods("call").satisfies(cl -> {
+            Method callMethod = cl.getMethod("call", List.class);
+            assertThat(callMethod.getAnnotation(Other.class)).isNull();
+
+            AnnotatedType annotatedType = callMethod.getAnnotatedParameterTypes()[0];
+            assertThat(annotatedType).isInstanceOfSatisfying(AnnotatedParameterizedType.class, apt -> {
+                AnnotatedType[] typeArguments = apt.getAnnotatedActualTypeArguments();
+                assertThat(typeArguments).singleElement().satisfies(t -> {
+                    assertThat(t.getAnnotation(Foo.class)).isNotNull();
+                });
+            });
+        });
+
+        assertThat(ClientProxy.unwrap(client).getClass()).hasMethods("call2").satisfies(cl -> {
+            Method callMethod = cl.getMethod("call2", String.class, Map.class);
+            assertThat(callMethod.getAnnotation(Other.class)).isNotNull();
+
+            AnnotatedType annotatedType1 = callMethod.getAnnotatedParameterTypes()[0];
+            assertThat(annotatedType1.getAnnotation(Bar.class)).isNotNull();
+            assertThat(annotatedType1.getAnnotation(Foo.class)).isNull();
+
+            AnnotatedType annotatedType2 = callMethod.getAnnotatedParameterTypes()[1];
+            assertThat(annotatedType2).isInstanceOfSatisfying(AnnotatedParameterizedType.class, apt -> {
+                AnnotatedType[] typeArguments = apt.getAnnotatedActualTypeArguments();
+                assertThat(typeArguments).hasSize(2).satisfies(t -> {
+                    AnnotatedType first = typeArguments[0];
+                    assertThat(first.getAnnotation(Foo.class)).isNotNull();
+                    assertThat(first.getAnnotation(Bar.class)).isNull();
+
+                    AnnotatedType second = typeArguments[1];
+                    assertThat(second.getAnnotation(Bar.class)).isNotNull();
+                    assertThat(second.getAnnotation(Foo.class)).isNull();
+                });
+            });
+        });
+    }
+
+    @RegisterRestClient(configKey = "client")
+    public interface Client {
+
+        @Path("/")
+        @POST
+        void call(List<@Foo Pojo> list);
+
+        @Other
+        @Path("/2")
+        @POST
+        void call2(@NotBody @Bar String p1, @NotBody Map<@Foo String, @Bar String> p2);
+    }
+
+    public record Pojo(String field) {
+
+    }
+
+    @Target({ ElementType.TYPE_USE, ElementType.PARAMETER })
+    @Retention(RetentionPolicy.RUNTIME)
+    public @interface Foo {
+    }
+
+    @Target({ ElementType.TYPE_USE })
+    @Retention(RetentionPolicy.RUNTIME)
+    public @interface Bar {
+    }
+
+    @Target({ ElementType.METHOD })
+    @Retention(RetentionPolicy.RUNTIME)
+    public @interface Other {
+
+    }
+}

